tests/test_discover.py

```diff
@@ -0,0 +1,195 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import annotations
+
+from os import mkdir
+from os.path import join, normpath
+from tempfile import TemporaryDirectory
+from typing import Dict, Union
+from unittest import TestCase
+
+from moz.l10n.paths import L10nDiscoverPaths, MissingSourceDirectoryError
+
+Tree = Dict[str, Union[str, "Tree"]]
+
+
+def build_file_tree(root: str, tree: Tree) -> None:
+    for name, value in tree.items():
+        path = join(root, name)
+        if isinstance(value, str):
+            with open(path, "x") as file:
+                if value:
+                    file.write(value)
+        else:
+            mkdir(path)
+            build_file_tree(path, value)
+
+
+class TestL10nDiscover(TestCase):
+    def test_not_found(self):
+        tree: Tree = {
+            "one.pot": "",
+            "two": {"a.ftl": "", "b.pot": ""},
+            "three": {"c": "", "d": {"e": ""}, "f": {"g.ftl": ""}},
+        }
+        with TemporaryDirectory() as root:
+            build_file_tree(root, tree)
+            with self.assertRaises(MissingSourceDirectoryError):
+                L10nDiscoverPaths(root)
+
+    def test_ref_only(self):
+        tree: Tree = {
+            "en": {
+                "one.pot": "",
+                "two": {"a.ftl": "", "b.pot": ""},
+                "three": {"c": "", "d": {"e": ""}, "f": {"g.ftl": ""}},
+            },
+        }
+        with TemporaryDirectory() as root:
+            build_file_tree(root, tree)
+            paths = L10nDiscoverPaths(root)
+
+        assert paths.ref_root == join(root, "en")
+        assert paths.base is None
+        assert paths.locales is None
+        with self.assertRaises(ValueError):
+            paths.all()
+        with self.assertRaises(ValueError):
+            paths.target_path("one.pot")
+        with self.assertRaises(ValueError):
+            paths.format_target_path("one.pot", "xx")
+
+        paths.base = join(root, "target")
+        expected = {
+            (
+                join(root, "en", normpath(ref)),
+                join(root, "target", "{locale}", normpath(tgt)),
+            ): None
+            for ref, tgt in {
+                "one.pot": "one.po",
+                "three/c": "three/c",
+                "three/d/e": "three/d/e",
+                "three/f/g.ftl": "three/f/g.ftl",
+                "two/a.ftl": "two/a.ftl",
+                "two/b.pot": "two/b.po",
+            }.items()
+        }
+        assert paths.all() == expected
+        for ref, tgt in expected:
+            assert paths.target_path(ref) == tgt
+            assert paths.target_path(ref, locale="xx") == tgt.format(locale="xx")
+
+    def test_ref_priorities(self):
```

Comment:
To actually test priority order, multiple reference directories should co-exist.

Reply:
They do co-exist here; each of the `build_file_tree()` commands is cumulative on the previous, as they're done within the same `with TemporaryDirectory()` block.